docs/source-fabric/guide/checkpoint.rst

```diff
@@ -123,6 +123,36 @@ Here is a trivial example to illustrate how it works:
 See also: `Saving and loading models in PyTorch <https://pytorch.org/tutorials/beginner/saving_loading_models.html>`_.
 
 
+----
+
+*************************
+Save a partial checkpoint
+*************************
+
+When saving a checkpoint using Fabric, you have the flexibility to choose which parameters to include in the saved file.
+This can be useful in scenarios such as fine-tuning, where you only want to save a subset of the parameters, reducing
+the size of the checkpoint and saving disk space.
+
+To accomplish this, you can use a filter during the saving proces. The filter is a function that determines whether
+an item should be saved (returning ``True``) or excluded (returning ``False``).
+The filter operates on dictionary objects and evaluates each key-value pair individually.
+
+Here's an example of using a filter when saving a checkpoint:
+
+.. code-block:: python
+
+    state = {"model": model, "optimizer": optimizer, "foo": 123}
+
+    # Apply the same filter to all keys in the state
+    filter = lambda k, v: "weight" in k or "param_groups" in k
+    fabric.save("path/to/checkpoint.ckpt", state, filter=filter)
+    # This will save {"model": {"layer.weight": ...}, "optimizer": {"param_groups": ...}, "foo": 123}
+
+    # Alternatively, specify a filter per key
+    filters = {"model": lambda k, v: "weight" in k, "optimizer": lambda k, v: "param_groups" in k}
```

Comment:
Filtering the optimizer state this way is not going to be useful. Parameters in most optimizers are referenced by value, not by name. In the issue #15481 I thought about excluding them automatically based on the params excluded in the model. But I think we should just not implement this right now and leave it for later.

So my recommendation is to only implement filter for the model params.

Reply:
The current implementation treats all `state_dict`s equally. This means the user could currently implement such behaviour for the optimizer state.

Doing it automatically would be nice sugar, but at least the current implementation unblocks those who want this. If we only apply the filter to nn.Modules then this is not possible.

Also it would not allow filtering other dict-like states

Reply:
Whether `nn.Module.state_dict()` and `Optimizer.state_dict()` gets treated the same way can be discussed yes. I think what we do should be guided by practical experience (e.g. in lit-llama/parrot). The way it is currently implemented in this PR would be quite cumbersome IMO (I can explain offline).